// @google-analytics/admin/build/src/v1beta/analytics_admin_service_client.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnalyticsAdminServiceClient = void 0;
const jsonProtos = require("../../protos/protos.json");
const google_gax_1 = require("google-gax");
/**
 * Client JSON configuration object, loaded from
 * `src/v1beta/analytics_admin_service_client_config.json`.
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./analytics_admin_service_client_config.json");
const version = require('../../../package.json').version;
/**
 *  Service Interface for the Google Analytics Admin API.
 * @class
 * @memberof v1beta
 */
class AnalyticsAdminServiceClient {
    _terminated = false;
    _opts;
    _providedCustomServicePath;
    _gaxModule;
    _gaxGrpc;
    _protos;
    _defaults;
    _universeDomain;
    _servicePath;
    _log = google_gax_1.loggingUtils.log('admin');
    auth;
    descriptors = {
        page: {},
        stream: {},
        longrunning: {},
        batching: {},
    };
    warn;
    innerApiCalls;
    pathTemplates;
    analyticsAdminServiceStub;
    /**
     * Construct an instance of AnalyticsAdminServiceClient.
     *
     * @param {object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param {object} [options.credentials] - Credentials object.
     * @param {string} [options.credentials.client_email]
     * @param {string} [options.credentials.private_key]
     * @param {string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param {string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param {number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param {string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     {@link https://cloud.google.com/docs/authentication/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param {string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param {gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of {@link gapicConfig}.
     * @param {boolean} [options.fallback] - Use HTTP/1.1 REST mode.
     *     For more information, please check the
     *     {@link https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#http11-rest-api-mode documentation}.
     * @param {gax} [gaxInstance]: loaded instance of `google-gax`. Useful if you
     *     need to avoid loading the default gRPC version and want to use the fallback
     *     HTTP implementation. Load only fallback version and pass it to the constructor:
     *     ```
     *     const gax = require('google-gax/build/src/fallback'); // avoids loading google-gax with gRPC
     *     const client = new AnalyticsAdminServiceClient({fallback: true}, gax);
     *     ```
     */
    constructor(opts, gaxInstance) {
        // Ensure that options include all the required fields.
        const staticMembers = this.constructor;
        if (opts?.universe_domain && opts?.universeDomain && opts?.universe_domain !== opts?.universeDomain) {
            throw new Error('Please set either universe_domain or universeDomain, but not both.');
        }
        const universeDomainEnvVar = (typeof process === 'object' && typeof process.env === 'object') ? process.env['GOOGLE_CLOUD_UNIVERSE_DOMAIN'] : undefined;
        this._universeDomain = opts?.universeDomain ?? opts?.universe_domain ?? universeDomainEnvVar ?? 'googleapis.com';
        this._servicePath = 'analyticsadmin.' + this._universeDomain;
        const servicePath = opts?.servicePath || opts?.apiEndpoint || this._servicePath;
        this._providedCustomServicePath = !!(opts?.servicePath || opts?.apiEndpoint);
        const port = opts?.port || staticMembers.port;
        const clientConfig = opts?.clientConfig ?? {};
        const fallback = opts?.fallback ?? (typeof window !== 'undefined' && typeof window?.fetch === 'function');
        opts = Object.assign({ servicePath, port, clientConfig, fallback }, opts);
        // Request numeric enum values if REST transport is used.
        opts.numericEnums = true;
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== this._servicePath && !('scopes' in opts)) {
            opts['scopes'] = staticMembers.scopes;
        }
        // Load google-gax module synchronously if needed
        if (!gaxInstance) {
            gaxInstance = require('google-gax');
        }
        // Choose either gRPC or proto-over-HTTP implementation of google-gax.
        this._gaxModule = opts.fallback ? gaxInstance.fallback : gaxInstance;
        // Create a `gaxGrpc` object, with any grpc-specific options sent to the client.
        this._gaxGrpc = new this._gaxModule.GrpcClient(opts);
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = this._gaxGrpc.auth;
        // Set useJWTAccessWithScope on the auth object.
        this.auth.useJWTAccessWithScope = true;
        // Set defaultServicePath on the auth object.
        this.auth.defaultServicePath = this._servicePath;
        // Set the default scopes in auth client if needed.
        if (servicePath === this._servicePath) {
            this.auth.defaultScopes = staticMembers.scopes;
        }
        // Determine the client header string.
        const clientHeader = [
            `gax/${this._gaxModule.version}`,
            `gapic/${version}`,
        ];
        if (typeof process === 'object' && 'versions' in process) {
            clientHeader.push(`gl-node/${process.versions.node}`);
        }
        else {
            clientHeader.push(`gl-web/${this._gaxModule.version}`);
        }
        if (!opts.fallback) {
            clientHeader.push(`grpc/${this._gaxGrpc.grpcVersion}`);
        }
        else {
            clientHeader.push(`rest/${this._gaxGrpc.grpcVersion}`);
        }
        if (opts.libName && opts.libVersion) {
            clientHeader.push(`${opts.libName}/${opts.libVersion}`);
        }
        // Load the applicable protos.
        this._protos = this._gaxGrpc.loadProtoJSON(jsonProtos);
        // This API contains "path templates"; forward-slash-separated
        // identifiers to uniquely identify resources within the API.
        // Create useful helper objects for these.
        this.pathTemplates = {
            accountPathTemplate: new this._gaxModule.PathTemplate('accounts/{account}'),
            accountSummaryPathTemplate: new this._gaxModule.PathTemplate('accountSummaries/{account_summary}'),
            conversionEventPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/conversionEvents/{conversion_event}'),
            customDimensionPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/customDimensions/{custom_dimension}'),
            customMetricPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/customMetrics/{custom_metric}'),
            dataRetentionSettingsPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataRetentionSettings'),
            dataSharingSettingsPathTemplate: new this._gaxModule.PathTemplate('accounts/{account}/dataSharingSettings'),
            dataStreamPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataStreams/{data_stream}'),
            firebaseLinkPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/firebaseLinks/{firebase_link}'),
            googleAdsLinkPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/googleAdsLinks/{google_ads_link}'),
            keyEventPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/keyEvents/{key_event}'),
            measurementProtocolSecretPathTemplate: new this._gaxModule.PathTemplate('properties/{property}/dataStreams/{data_stream}/measurementProtocolSecrets/{measurement_protocol_secret}'),
            propertyPathTemplate: new this._gaxModule.PathTemplate('properties/{property}'),
        };
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = {
            listAccounts: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'accounts'),
            listAccountSummaries: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'accountSummaries'),
            listProperties: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'properties'),
            listFirebaseLinks: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'firebaseLinks'),
            listGoogleAdsLinks: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'googleAdsLinks'),
            listMeasurementProtocolSecrets: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'measurementProtocolSecrets'),
            searchChangeHistoryEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'changeHistoryEvents'),
            listConversionEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'conversionEvents'),
            listKeyEvents: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'keyEvents'),
            listCustomDimensions: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'customDimensions'),
            listCustomMetrics: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'customMetrics'),
            listDataStreams: new this._gaxModule.PageDescriptor('pageToken', 'nextPageToken', 'dataStreams')
        };
        // Put together the default options sent with requests.
        this._defaults = this._gaxGrpc.constructSettings('google.analytics.admin.v1beta.AnalyticsAdminService', gapicConfig, opts.clientConfig || {}, { 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = {};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = this._gaxModule.warn;
    }
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns {Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() {
        // If the client stub promise is already initialized, return immediately.
        if (this.analyticsAdminServiceStub) {
            return this.analyticsAdminServiceStub;
        }
        // Put together the "service stub" for
        // google.analytics.admin.v1beta.AnalyticsAdminService.
        this.analyticsAdminServiceStub = this._gaxGrpc.createStub(this._opts.fallback ?
            this._protos.lookupService('google.analytics.admin.v1beta.AnalyticsAdminService') :
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            this._protos.google.analytics.admin.v1beta.AnalyticsAdminService, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const analyticsAdminServiceStubMethods = ['getAccount', 'listAccounts', 'deleteAccount', 'updateAccount', 'provisionAccountTicket', 'listAccountSummaries', 'getProperty', 'listProperties', 'createProperty', 'deleteProperty', 'updateProperty', 'createFirebaseLink', 'deleteFirebaseLink', 'listFirebaseLinks', 'createGoogleAdsLink', 'updateGoogleAdsLink', 'deleteGoogleAdsLink', 'listGoogleAdsLinks', 'getDataSharingSettings', 'getMeasurementProtocolSecret', 'listMeasurementProtocolSecrets', 'createMeasurementProtocolSecret', 'deleteMeasurementProtocolSecret', 'updateMeasurementProtocolSecret', 'acknowledgeUserDataCollection', 'searchChangeHistoryEvents', 'createConversionEvent', 'updateConversionEvent', 'getConversionEvent', 'deleteConversionEvent', 'listConversionEvents', 'createKeyEvent', 'updateKeyEvent', 'getKeyEvent', 'deleteKeyEvent', 'listKeyEvents', 'createCustomDimension', 'updateCustomDimension', 'listCustomDimensions', 'archiveCustomDimension', 'getCustomDimension', 'createCustomMetric', 'updateCustomMetric', 'listCustomMetrics', 'archiveCustomMetric', 'getCustomMetric', 'getDataRetentionSettings', 'updateDataRetentionSettings', 'createDataStream', 'deleteDataStream', 'updateDataStream', 'listDataStreams', 'getDataStream', 'runAccessReport'];
        for (const methodName of analyticsAdminServiceStubMethods) {
            const callPromise = this.analyticsAdminServiceStub.then(stub => (...args) => {
                if (this._terminated) {
                    return Promise.reject('The client has already been closed.');
                }
                const func = stub[methodName];
                return func.apply(stub, args);
            }, (err) => () => {
                throw err;
            });
            const descriptor = this.descriptors.page[methodName] ||
                undefined;
            const apiCall = this._gaxModule.createApiCall(callPromise, this._defaults[methodName], descriptor, this._opts.fallback);
            this.innerApiCalls[methodName] = apiCall;
        }
        return this.analyticsAdminServiceStub;
    }
    /**
     * The DNS address for this API service.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get servicePath() {
        if (typeof process === 'object' && typeof process.emitWarning === 'function') {
            process.emitWarning('Static servicePath is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'analyticsadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service - same as servicePath.
     * @deprecated Use the apiEndpoint method of the client instance.
     * @returns {string} The DNS address for this service.
     */
    static get apiEndpoint() {
        if (typeof process === 'object' && typeof process.emitWarning === 'function') {
            process.emitWarning('Static apiEndpoint is deprecated, please use the instance method instead.', 'DeprecationWarning');
        }
        return 'analyticsadmin.googleapis.com';
    }
    /**
     * The DNS address for this API service.
     * @returns {string} The DNS address for this service.
     */
    get apiEndpoint() {
        return this._servicePath;
    }
    get universeDomain() {
        return this._universeDomain;
    }
    /**
     * The port for this API service.
     * @returns {number} The default port for this service.
     */
    static get port() {
        return 443;
    }
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns {string[]} List of default scopes.
     */
    static get scopes() {
        return [
            'https://www.googleapis.com/auth/analytics.edit',
            'https://www.googleapis.com/auth/analytics.readonly'
        ];
    }
    /**
     * Return the project ID used by this class.
     * @returns {Promise} A promise that resolves to string containing the project ID.
     */
    getProjectId(callback) {
        if (callback) {
            this.auth.getProjectId(callback);
            return;
        }
        return this.auth.getProjectId();
    }
    getAccount(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getAccount request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getAccount response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getAccount(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getAccount response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteAccount(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteAccount request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteAccount response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteAccount(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteAccount response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateAccount(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'account.name': request.account.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateAccount request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateAccount response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateAccount(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateAccount response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    provisionAccountTicket(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize().catch(err => { throw err; });
        this._log.info('provisionAccountTicket request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('provisionAccountTicket response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.provisionAccountTicket(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('provisionAccountTicket response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getProperty(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getProperty request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getProperty response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getProperty(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getProperty response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createProperty(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize().catch(err => { throw err; });
        this._log.info('createProperty request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createProperty response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createProperty(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createProperty response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteProperty(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteProperty request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteProperty response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteProperty(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteProperty response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateProperty(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'property.name': request.property.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateProperty request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateProperty response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateProperty(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateProperty response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createFirebaseLink(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createFirebaseLink request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createFirebaseLink response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createFirebaseLink(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createFirebaseLink response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteFirebaseLink(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteFirebaseLink request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteFirebaseLink response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteFirebaseLink(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteFirebaseLink response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createGoogleAdsLink(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createGoogleAdsLink request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createGoogleAdsLink response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createGoogleAdsLink(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createGoogleAdsLink response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateGoogleAdsLink(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'google_ads_link.name': request.googleAdsLink.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateGoogleAdsLink request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateGoogleAdsLink response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateGoogleAdsLink(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateGoogleAdsLink response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteGoogleAdsLink(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteGoogleAdsLink request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteGoogleAdsLink response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteGoogleAdsLink(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteGoogleAdsLink response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getDataSharingSettings(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getDataSharingSettings request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getDataSharingSettings response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getDataSharingSettings(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getDataSharingSettings response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getMeasurementProtocolSecret request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getMeasurementProtocolSecret response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getMeasurementProtocolSecret(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getMeasurementProtocolSecret response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createMeasurementProtocolSecret request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createMeasurementProtocolSecret response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createMeasurementProtocolSecret(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createMeasurementProtocolSecret response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteMeasurementProtocolSecret request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteMeasurementProtocolSecret response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteMeasurementProtocolSecret(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteMeasurementProtocolSecret response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateMeasurementProtocolSecret(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'measurement_protocol_secret.name': request.measurementProtocolSecret.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateMeasurementProtocolSecret request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateMeasurementProtocolSecret response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateMeasurementProtocolSecret(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateMeasurementProtocolSecret response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    acknowledgeUserDataCollection(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'property': request.property ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('acknowledgeUserDataCollection request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('acknowledgeUserDataCollection response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.acknowledgeUserDataCollection(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('acknowledgeUserDataCollection response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createConversionEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$CreateConversionEvent', 'CreateConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('createConversionEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createConversionEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createConversionEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createConversionEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateConversionEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'conversion_event.name': request.conversionEvent.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$UpdateConversionEvent', 'UpdateConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('updateConversionEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateConversionEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateConversionEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateConversionEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getConversionEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$GetConversionEvent', 'GetConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('getConversionEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getConversionEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getConversionEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getConversionEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteConversionEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$DeleteConversionEvent', 'DeleteConversionEvent is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('deleteConversionEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteConversionEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteConversionEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteConversionEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createKeyEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createKeyEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createKeyEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createKeyEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createKeyEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateKeyEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'key_event.name': request.keyEvent.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateKeyEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateKeyEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateKeyEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateKeyEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getKeyEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getKeyEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getKeyEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getKeyEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getKeyEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteKeyEvent(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteKeyEvent request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteKeyEvent response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteKeyEvent(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteKeyEvent response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createCustomDimension(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createCustomDimension request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createCustomDimension response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createCustomDimension(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createCustomDimension response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateCustomDimension(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'custom_dimension.name': request.customDimension.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateCustomDimension request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateCustomDimension response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateCustomDimension(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateCustomDimension response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    archiveCustomDimension(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('archiveCustomDimension request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('archiveCustomDimension response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.archiveCustomDimension(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('archiveCustomDimension response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getCustomDimension(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getCustomDimension request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getCustomDimension response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getCustomDimension(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getCustomDimension response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createCustomMetric(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createCustomMetric request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createCustomMetric response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createCustomMetric(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createCustomMetric response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateCustomMetric(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'custom_metric.name': request.customMetric.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateCustomMetric request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateCustomMetric response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateCustomMetric(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateCustomMetric response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    archiveCustomMetric(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('archiveCustomMetric request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('archiveCustomMetric response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.archiveCustomMetric(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('archiveCustomMetric response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getCustomMetric(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getCustomMetric request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getCustomMetric response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getCustomMetric(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getCustomMetric response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getDataRetentionSettings(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getDataRetentionSettings request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getDataRetentionSettings response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getDataRetentionSettings(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getDataRetentionSettings response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateDataRetentionSettings(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'data_retention_settings.name': request.dataRetentionSettings.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateDataRetentionSettings request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateDataRetentionSettings response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateDataRetentionSettings(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateDataRetentionSettings response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    createDataStream(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('createDataStream request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('createDataStream response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.createDataStream(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('createDataStream response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    deleteDataStream(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('deleteDataStream request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('deleteDataStream response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.deleteDataStream(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('deleteDataStream response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    updateDataStream(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'data_stream.name': request.dataStream.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('updateDataStream request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('updateDataStream response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.updateDataStream(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('updateDataStream response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    getDataStream(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'name': request.name ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('getDataStream request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('getDataStream response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.getDataStream(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('getDataStream response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    runAccessReport(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'entity': request.entity ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this._log.info('runAccessReport request %j', request);
        const wrappedCallback = callback
            ? (error, response, options, rawResponse) => {
                this._log.info('runAccessReport response %j', response);
                callback(error, response, options, rawResponse); // We verified callback above.
            }
            : undefined;
        return this.innerApiCalls.runAccessReport(request, options, wrappedCallback)
            ?.then(([response, options, rawResponse]) => {
            this._log.info('runAccessReport response %j', response);
            return [response, options, rawResponse];
        }).catch((error) => {
            if (error && 'statusDetails' in error && error.statusDetails instanceof Array) {
                const protos = this._gaxModule.protobuf.Root.fromJSON(jsonProtos);
                error.statusDetails = (0, google_gax_1.decodeAnyProtosInArray)(error.statusDetails, protos);
            }
            throw error;
        });
    }
    listAccounts(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listAccounts values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listAccounts request %j', request);
        return this.innerApiCalls
            .listAccounts(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listAccounts values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listAccounts`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccounts` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccounts` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Accounts in the
     *   results. Accounts can be inspected to determine whether they are deleted or
     *   not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.Account|Account} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listAccountsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listAccountsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccounts'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listAccounts stream %j', request);
        return this.descriptors.page.listAccounts.createStream(this.innerApiCalls.listAccounts, request, callSettings);
    }
    /**
     * Equivalent to `listAccounts`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccounts` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccounts` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Accounts in the
     *   results. Accounts can be inspected to determine whether they are deleted or
     *   not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.Account|Account}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_accounts.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListAccounts_async
     */
    listAccountsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccounts'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listAccounts iterate %j', request);
        return this.descriptors.page.listAccounts.asyncIterate(this.innerApiCalls['listAccounts'], request, callSettings);
    }
    listAccountSummaries(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listAccountSummaries values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listAccountSummaries request %j', request);
        return this.innerApiCalls
            .listAccountSummaries(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listAccountSummaries values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listAccountSummaries`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of AccountSummary resources to return. The service may
     *   return fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccountSummaries` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccountSummaries`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.AccountSummary|AccountSummary} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listAccountSummariesAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listAccountSummariesStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccountSummaries'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listAccountSummaries stream %j', request);
        return this.descriptors.page.listAccountSummaries.createStream(this.innerApiCalls.listAccountSummaries, request, callSettings);
    }
    /**
     * Equivalent to `listAccountSummaries`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {number} request.pageSize
     *   The maximum number of AccountSummary resources to return. The service may
     *   return fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListAccountSummaries` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListAccountSummaries`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.AccountSummary|AccountSummary}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_account_summaries.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListAccountSummaries_async
     */
    listAccountSummariesAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listAccountSummaries'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listAccountSummaries iterate %j', request);
        return this.descriptors.page.listAccountSummaries.asyncIterate(this.innerApiCalls['listAccountSummaries'], request, callSettings);
    }
    listProperties(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listProperties values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listProperties request %j', request);
        return this.innerApiCalls
            .listProperties(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listProperties values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listProperties`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.filter
     *   Required. An expression for filtering the results of the request.
     *   Fields eligible for filtering are:
     *   `parent:`(The resource name of the parent account/property) or
     *   `ancestor:`(The resource name of the parent account) or
     *   `firebase_project:`(The id or number of the linked firebase project).
     *   Some examples of filters:
     *
     *   ```
     *   | Filter                      | Description                               |
     *   |-----------------------------|-------------------------------------------|
     *   | parent:accounts/123         | The account with account id: 123.       |
     *   | parent:properties/123       | The property with property id: 123.       |
     *   | ancestor:accounts/123       | The account with account id: 123.         |
     *   | firebase_project:project-id | The firebase project with id: project-id. |
     *   | firebase_project:123        | The firebase project with number: 123.    |
     *   ```
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListProperties` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListProperties` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Properties in the
     *   results. Properties can be inspected to determine whether they are deleted
     *   or not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.Property|Property} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listPropertiesAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listPropertiesStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listProperties'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listProperties stream %j', request);
        return this.descriptors.page.listProperties.createStream(this.innerApiCalls.listProperties, request, callSettings);
    }
    /**
     * Equivalent to `listProperties`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.filter
     *   Required. An expression for filtering the results of the request.
     *   Fields eligible for filtering are:
     *   `parent:`(The resource name of the parent account/property) or
     *   `ancestor:`(The resource name of the parent account) or
     *   `firebase_project:`(The id or number of the linked firebase project).
     *   Some examples of filters:
     *
     *   ```
     *   | Filter                      | Description                               |
     *   |-----------------------------|-------------------------------------------|
     *   | parent:accounts/123         | The account with account id: 123.       |
     *   | parent:properties/123       | The property with property id: 123.       |
     *   | ancestor:accounts/123       | The account with account id: 123.         |
     *   | firebase_project:project-id | The firebase project with id: project-id. |
     *   | firebase_project:123        | The firebase project with number: 123.    |
     *   ```
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListProperties` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListProperties` must
     *   match the call that provided the page token.
     * @param {boolean} request.showDeleted
     *   Whether to include soft-deleted (ie: "trashed") Properties in the
     *   results. Properties can be inspected to determine whether they are deleted
     *   or not.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.Property|Property}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_properties.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListProperties_async
     */
    listPropertiesAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        const defaultCallSettings = this._defaults['listProperties'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listProperties iterate %j', request);
        return this.descriptors.page.listProperties.asyncIterate(this.innerApiCalls['listProperties'], request, callSettings);
    }
    listFirebaseLinks(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listFirebaseLinks values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listFirebaseLinks request %j', request);
        return this.innerApiCalls
            .listFirebaseLinks(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listFirebaseLinks values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listFirebaseLinks`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Format: properties/{property_id}
     *
     *   Example: `properties/1234`
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListFirebaseLinks` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListFirebaseLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.FirebaseLink|FirebaseLink} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listFirebaseLinksAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listFirebaseLinksStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listFirebaseLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listFirebaseLinks stream %j', request);
        return this.descriptors.page.listFirebaseLinks.createStream(this.innerApiCalls.listFirebaseLinks, request, callSettings);
    }
    /**
     * Equivalent to `listFirebaseLinks`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Format: properties/{property_id}
     *
     *   Example: `properties/1234`
     * @param {number} request.pageSize
     *   The maximum number of resources to return. The service may return
     *   fewer than this value, even if there are additional pages.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListFirebaseLinks` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListFirebaseLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.FirebaseLink|FirebaseLink}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_firebase_links.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListFirebaseLinks_async
     */
    listFirebaseLinksAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listFirebaseLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listFirebaseLinks iterate %j', request);
        return this.descriptors.page.listFirebaseLinks.asyncIterate(this.innerApiCalls['listFirebaseLinks'], request, callSettings);
    }
    listGoogleAdsLinks(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listGoogleAdsLinks values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listGoogleAdsLinks request %j', request);
        return this.innerApiCalls
            .listGoogleAdsLinks(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listGoogleAdsLinks values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listGoogleAdsLinks`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListGoogleAdsLinks` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListGoogleAdsLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.GoogleAdsLink|GoogleAdsLink} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listGoogleAdsLinksAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listGoogleAdsLinksStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listGoogleAdsLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listGoogleAdsLinks stream %j', request);
        return this.descriptors.page.listGoogleAdsLinks.createStream(this.innerApiCalls.listGoogleAdsLinks, request, callSettings);
    }
    /**
     * Equivalent to `listGoogleAdsLinks`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListGoogleAdsLinks` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListGoogleAdsLinks` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.GoogleAdsLink|GoogleAdsLink}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_google_ads_links.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListGoogleAdsLinks_async
     */
    listGoogleAdsLinksAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listGoogleAdsLinks'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listGoogleAdsLinks iterate %j', request);
        return this.descriptors.page.listGoogleAdsLinks.asyncIterate(this.innerApiCalls['listGoogleAdsLinks'], request, callSettings);
    }
    listMeasurementProtocolSecrets(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listMeasurementProtocolSecrets values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listMeasurementProtocolSecrets request %j', request);
        return this.innerApiCalls
            .listMeasurementProtocolSecrets(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listMeasurementProtocolSecrets values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listMeasurementProtocolSecrets`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent stream.
     *   Format:
     *   properties/{property}/dataStreams/{dataStream}/measurementProtocolSecrets
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 10 resources will be returned.
     *   The maximum value is 10. Higher values will be coerced to the maximum.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListMeasurementProtocolSecrets`
     *   call. Provide this to retrieve the subsequent page. When paginating, all
     *   other parameters provided to `ListMeasurementProtocolSecrets` must match
     *   the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.MeasurementProtocolSecret|MeasurementProtocolSecret} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listMeasurementProtocolSecretsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listMeasurementProtocolSecretsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listMeasurementProtocolSecrets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listMeasurementProtocolSecrets stream %j', request);
        return this.descriptors.page.listMeasurementProtocolSecrets.createStream(this.innerApiCalls.listMeasurementProtocolSecrets, request, callSettings);
    }
    /**
     * Equivalent to `listMeasurementProtocolSecrets`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent stream.
     *   Format:
     *   properties/{property}/dataStreams/{dataStream}/measurementProtocolSecrets
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 10 resources will be returned.
     *   The maximum value is 10. Higher values will be coerced to the maximum.
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListMeasurementProtocolSecrets`
     *   call. Provide this to retrieve the subsequent page. When paginating, all
     *   other parameters provided to `ListMeasurementProtocolSecrets` must match
     *   the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.MeasurementProtocolSecret|MeasurementProtocolSecret}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_measurement_protocol_secrets.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListMeasurementProtocolSecrets_async
     */
    listMeasurementProtocolSecretsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listMeasurementProtocolSecrets'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listMeasurementProtocolSecrets iterate %j', request);
        return this.descriptors.page.listMeasurementProtocolSecrets.asyncIterate(this.innerApiCalls['listMeasurementProtocolSecrets'], request, callSettings);
    }
    searchChangeHistoryEvents(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'account': request.account ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('searchChangeHistoryEvents values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('searchChangeHistoryEvents request %j', request);
        return this.innerApiCalls
            .searchChangeHistoryEvents(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('searchChangeHistoryEvents values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `searchChangeHistoryEvents`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.account
     *   Required. The account resource for which to return change history
     *   resources. Format: accounts/{account}
     *
     *   Example: `accounts/100`
     * @param {string} [request.property]
     *   Optional. Resource name for a child property. If set, only return changes
     *   made to this property or its child resources.
     *   Format: properties/{propertyId}
     *
     *   Example: `properties/100`
     * @param {number[]} [request.resourceType]
     *   Optional. If set, only return changes if they are for a resource that
     *   matches at least one of these types.
     * @param {number[]} [request.action]
     *   Optional. If set, only return changes that match one or more of these types
     *   of actions.
     * @param {string[]} [request.actorEmail]
     *   Optional. If set, only return changes if they are made by a user in this
     *   list.
     * @param {google.protobuf.Timestamp} [request.earliestChangeTime]
     *   Optional. If set, only return changes made after this time (inclusive).
     * @param {google.protobuf.Timestamp} [request.latestChangeTime]
     *   Optional. If set, only return changes made before this time (inclusive).
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of ChangeHistoryEvent items to return.
     *   If unspecified, at most 50 items will be returned. The maximum value is 200
     *   (higher values will be coerced to the maximum).
     *
     *   Note that the service may return a page with fewer items than this value
     *   specifies (potentially even zero), and that there still may be additional
     *   pages. If you want a particular number of items, you'll need to continue
     *   requesting additional pages using `page_token` until you get the needed
     *   number.
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous
     *   `SearchChangeHistoryEvents` call. Provide this to retrieve the subsequent
     *   page. When paginating, all other parameters provided to
     *   `SearchChangeHistoryEvents` must match the call that provided the page
     *   token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.ChangeHistoryEvent|ChangeHistoryEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `searchChangeHistoryEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    searchChangeHistoryEventsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'account': request.account ?? '',
        });
        const defaultCallSettings = this._defaults['searchChangeHistoryEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('searchChangeHistoryEvents stream %j', request);
        return this.descriptors.page.searchChangeHistoryEvents.createStream(this.innerApiCalls.searchChangeHistoryEvents, request, callSettings);
    }
    /**
     * Equivalent to `searchChangeHistoryEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.account
     *   Required. The account resource for which to return change history
     *   resources. Format: accounts/{account}
     *
     *   Example: `accounts/100`
     * @param {string} [request.property]
     *   Optional. Resource name for a child property. If set, only return changes
     *   made to this property or its child resources.
     *   Format: properties/{propertyId}
     *
     *   Example: `properties/100`
     * @param {number[]} [request.resourceType]
     *   Optional. If set, only return changes if they are for a resource that
     *   matches at least one of these types.
     * @param {number[]} [request.action]
     *   Optional. If set, only return changes that match one or more of these types
     *   of actions.
     * @param {string[]} [request.actorEmail]
     *   Optional. If set, only return changes if they are made by a user in this
     *   list.
     * @param {google.protobuf.Timestamp} [request.earliestChangeTime]
     *   Optional. If set, only return changes made after this time (inclusive).
     * @param {google.protobuf.Timestamp} [request.latestChangeTime]
     *   Optional. If set, only return changes made before this time (inclusive).
     * @param {number} [request.pageSize]
     *   Optional. The maximum number of ChangeHistoryEvent items to return.
     *   If unspecified, at most 50 items will be returned. The maximum value is 200
     *   (higher values will be coerced to the maximum).
     *
     *   Note that the service may return a page with fewer items than this value
     *   specifies (potentially even zero), and that there still may be additional
     *   pages. If you want a particular number of items, you'll need to continue
     *   requesting additional pages using `page_token` until you get the needed
     *   number.
     * @param {string} [request.pageToken]
     *   Optional. A page token, received from a previous
     *   `SearchChangeHistoryEvents` call. Provide this to retrieve the subsequent
     *   page. When paginating, all other parameters provided to
     *   `SearchChangeHistoryEvents` must match the call that provided the page
     *   token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.ChangeHistoryEvent|ChangeHistoryEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.search_change_history_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_SearchChangeHistoryEvents_async
     */
    searchChangeHistoryEventsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'account': request.account ?? '',
        });
        const defaultCallSettings = this._defaults['searchChangeHistoryEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('searchChangeHistoryEvents iterate %j', request);
        return this.descriptors.page.searchChangeHistoryEvents.asyncIterate(this.innerApiCalls['searchChangeHistoryEvents'], request, callSettings);
    }
    listConversionEvents(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listConversionEvents values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listConversionEvents request %j', request);
        return this.innerApiCalls
            .listConversionEvents(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listConversionEvents values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listConversionEvents`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListConversionEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListConversionEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.ConversionEvent|ConversionEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listConversionEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @deprecated ListConversionEvents is deprecated and may be removed in a future version.
     */
    listConversionEventsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listConversionEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('listConversionEvents stream %j', request);
        return this.descriptors.page.listConversionEvents.createStream(this.innerApiCalls.listConversionEvents, request, callSettings);
    }
    /**
     * Equivalent to `listConversionEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListConversionEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListConversionEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.ConversionEvent|ConversionEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_conversion_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListConversionEvents_async
     * @deprecated ListConversionEvents is deprecated and may be removed in a future version.
     */
    listConversionEventsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listConversionEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this.warn('DEP$AnalyticsAdminService-$ListConversionEvents', 'ListConversionEvents is deprecated and may be removed in a future version.', 'DeprecationWarning');
        this._log.info('listConversionEvents iterate %j', request);
        return this.descriptors.page.listConversionEvents.asyncIterate(this.innerApiCalls['listConversionEvents'], request, callSettings);
    }
    listKeyEvents(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listKeyEvents values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listKeyEvents request %j', request);
        return this.innerApiCalls
            .listKeyEvents(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listKeyEvents values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listKeyEvents`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListKeyEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListKeyEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.KeyEvent|KeyEvent} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listKeyEventsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listKeyEventsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listKeyEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listKeyEvents stream %j', request);
        return this.descriptors.page.listKeyEvents.createStream(this.innerApiCalls.listKeyEvents, request, callSettings);
    }
    /**
     * Equivalent to `listKeyEvents`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. The resource name of the parent property.
     *   Example: 'properties/123'
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200; (higher values will be coerced to the maximum)
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListKeyEvents` call.
     *   Provide this to retrieve the subsequent page.
     *   When paginating, all other parameters provided to `ListKeyEvents`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.KeyEvent|KeyEvent}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_key_events.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListKeyEvents_async
     */
    listKeyEventsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listKeyEvents'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listKeyEvents iterate %j', request);
        return this.descriptors.page.listKeyEvents.asyncIterate(this.innerApiCalls['listKeyEvents'], request, callSettings);
    }
    listCustomDimensions(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listCustomDimensions values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listCustomDimensions request %j', request);
        return this.innerApiCalls
            .listCustomDimensions(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listCustomDimensions values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listCustomDimensions`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomDimensions` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomDimensions`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.CustomDimension|CustomDimension} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listCustomDimensionsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listCustomDimensionsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listCustomDimensions'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listCustomDimensions stream %j', request);
        return this.descriptors.page.listCustomDimensions.createStream(this.innerApiCalls.listCustomDimensions, request, callSettings);
    }
    /**
     * Equivalent to `listCustomDimensions`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomDimensions` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomDimensions`
     *   must match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.CustomDimension|CustomDimension}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_custom_dimensions.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListCustomDimensions_async
     */
    listCustomDimensionsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listCustomDimensions'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listCustomDimensions iterate %j', request);
        return this.descriptors.page.listCustomDimensions.asyncIterate(this.innerApiCalls['listCustomDimensions'], request, callSettings);
    }
    listCustomMetrics(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listCustomMetrics values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listCustomMetrics request %j', request);
        return this.innerApiCalls
            .listCustomMetrics(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listCustomMetrics values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listCustomMetrics`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomMetrics` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomMetrics` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.CustomMetric|CustomMetric} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listCustomMetricsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listCustomMetricsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listCustomMetrics'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listCustomMetrics stream %j', request);
        return this.descriptors.page.listCustomMetrics.createStream(this.innerApiCalls.listCustomMetrics, request, callSettings);
    }
    /**
     * Equivalent to `listCustomMetrics`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListCustomMetrics` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListCustomMetrics` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.CustomMetric|CustomMetric}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_custom_metrics.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListCustomMetrics_async
     */
    listCustomMetricsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listCustomMetrics'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listCustomMetrics iterate %j', request);
        return this.descriptors.page.listCustomMetrics.asyncIterate(this.innerApiCalls['listCustomMetrics'], request, callSettings);
    }
    listDataStreams(request, optionsOrCallback, callback) {
        request = request || {};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) {
            callback = optionsOrCallback;
            options = {};
        }
        else {
            options = optionsOrCallback;
        }
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        this.initialize().catch(err => { throw err; });
        const wrappedCallback = callback
            ? (error, values, nextPageRequest, rawResponse) => {
                this._log.info('listDataStreams values %j', values);
                callback(error, values, nextPageRequest, rawResponse); // We verified callback above.
            }
            : undefined;
        this._log.info('listDataStreams request %j', request);
        return this.innerApiCalls
            .listDataStreams(request, options, wrappedCallback)
            ?.then(([response, input, output]) => {
            this._log.info('listDataStreams values %j', response);
            return [response, input, output];
        });
    }
    /**
     * Equivalent to `listDataStreams`, but returns a NodeJS Stream object.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDataStreams` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDataStreams` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Stream}
     *   An object stream which emits an object representing {@link protos.google.analytics.admin.v1beta.DataStream|DataStream} on 'data' event.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed. Note that it can affect your quota.
     *   We recommend using `listDataStreamsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     */
    listDataStreamsStream(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listDataStreams'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listDataStreams stream %j', request);
        return this.descriptors.page.listDataStreams.createStream(this.innerApiCalls.listDataStreams, request, callSettings);
    }
    /**
     * Equivalent to `listDataStreams`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param {Object} request
     *   The request object that will be sent.
     * @param {string} request.parent
     *   Required. Example format: properties/1234
     * @param {number} request.pageSize
     *   The maximum number of resources to return.
     *   If unspecified, at most 50 resources will be returned.
     *   The maximum value is 200 (higher values will be coerced to the maximum).
     * @param {string} request.pageToken
     *   A page token, received from a previous `ListDataStreams` call.
     *   Provide this to retrieve the subsequent page.
     *
     *   When paginating, all other parameters provided to `ListDataStreams` must
     *   match the call that provided the page token.
     * @param {object} [options]
     *   Call options. See {@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns {Object}
     *   An iterable Object that allows {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols | async iteration }.
     *   When you iterate the returned iterable, each element will be an object representing
     *   {@link protos.google.analytics.admin.v1beta.DataStream|DataStream}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the {@link https://github.com/googleapis/gax-nodejs/blob/master/client-libraries.md#auto-pagination | documentation }
     *   for more details and examples.
     * @example <caption>include:samples/generated/v1beta/analytics_admin_service.list_data_streams.js</caption>
     * region_tag:analyticsadmin_v1beta_generated_AnalyticsAdminService_ListDataStreams_async
     */
    listDataStreamsAsync(request, options) {
        request = request || {};
        options = options || {};
        options.otherArgs = options.otherArgs || {};
        options.otherArgs.headers = options.otherArgs.headers || {};
        options.otherArgs.headers['x-goog-request-params'] = this._gaxModule.routingHeader.fromParams({
            'parent': request.parent ?? '',
        });
        const defaultCallSettings = this._defaults['listDataStreams'];
        const callSettings = defaultCallSettings.merge(options);
        this.initialize().catch(err => { throw err; });
        this._log.info('listDataStreams iterate %j', request);
        return this.descriptors.page.listDataStreams.asyncIterate(this.innerApiCalls['listDataStreams'], request, callSettings);
    }
    // --------------------
    // -- Path templates --
    // --------------------
    /**
     * Return a fully-qualified account resource name string.
     *
     * @param {string} account
     * @returns {string} Resource name string.
     */
    accountPath(account) {
        return this.pathTemplates.accountPathTemplate.render({
            account: account,
        });
    }
    /**
     * Parse the account from Account resource.
     *
     * @param {string} accountName
     *   A fully-qualified path representing Account resource.
     * @returns {string} A string representing the account.
     */
    matchAccountFromAccountName(accountName) {
        return this.pathTemplates.accountPathTemplate.match(accountName).account;
    }
    /**
     * Return a fully-qualified accountSummary resource name string.
     *
     * @param {string} account_summary
     * @returns {string} Resource name string.
     */
    accountSummaryPath(accountSummary) {
        return this.pathTemplates.accountSummaryPathTemplate.render({
            account_summary: accountSummary,
        });
    }
    /**
     * Parse the account_summary from AccountSummary resource.
     *
     * @param {string} accountSummaryName
     *   A fully-qualified path representing AccountSummary resource.
     * @returns {string} A string representing the account_summary.
     */
    matchAccountSummaryFromAccountSummaryName(accountSummaryName) {
        return this.pathTemplates.accountSummaryPathTemplate.match(accountSummaryName).account_summary;
    }
    /**
     * Return a fully-qualified conversionEvent resource name string.
     *
     * @param {string} property
     * @param {string} conversion_event
     * @returns {string} Resource name string.
     */
    conversionEventPath(property, conversionEvent) {
        return this.pathTemplates.conversionEventPathTemplate.render({
            property: property,
            conversion_event: conversionEvent,
        });
    }
    /**
     * Parse the property from ConversionEvent resource.
     *
     * @param {string} conversionEventName
     *   A fully-qualified path representing ConversionEvent resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromConversionEventName(conversionEventName) {
        return this.pathTemplates.conversionEventPathTemplate.match(conversionEventName).property;
    }
    /**
     * Parse the conversion_event from ConversionEvent resource.
     *
     * @param {string} conversionEventName
     *   A fully-qualified path representing ConversionEvent resource.
     * @returns {string} A string representing the conversion_event.
     */
    matchConversionEventFromConversionEventName(conversionEventName) {
        return this.pathTemplates.conversionEventPathTemplate.match(conversionEventName).conversion_event;
    }
    /**
     * Return a fully-qualified customDimension resource name string.
     *
     * @param {string} property
     * @param {string} custom_dimension
     * @returns {string} Resource name string.
     */
    customDimensionPath(property, customDimension) {
        return this.pathTemplates.customDimensionPathTemplate.render({
            property: property,
            custom_dimension: customDimension,
        });
    }
    /**
     * Parse the property from CustomDimension resource.
     *
     * @param {string} customDimensionName
     *   A fully-qualified path representing CustomDimension resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromCustomDimensionName(customDimensionName) {
        return this.pathTemplates.customDimensionPathTemplate.match(customDimensionName).property;
    }
    /**
     * Parse the custom_dimension from CustomDimension resource.
     *
     * @param {string} customDimensionName
     *   A fully-qualified path representing CustomDimension resource.
     * @returns {string} A string representing the custom_dimension.
     */
    matchCustomDimensionFromCustomDimensionName(customDimensionName) {
        return this.pathTemplates.customDimensionPathTemplate.match(customDimensionName).custom_dimension;
    }
    /**
     * Return a fully-qualified customMetric resource name string.
     *
     * @param {string} property
     * @param {string} custom_metric
     * @returns {string} Resource name string.
     */
    customMetricPath(property, customMetric) {
        return this.pathTemplates.customMetricPathTemplate.render({
            property: property,
            custom_metric: customMetric,
        });
    }
    /**
     * Parse the property from CustomMetric resource.
     *
     * @param {string} customMetricName
     *   A fully-qualified path representing CustomMetric resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromCustomMetricName(customMetricName) {
        return this.pathTemplates.customMetricPathTemplate.match(customMetricName).property;
    }
    /**
     * Parse the custom_metric from CustomMetric resource.
     *
     * @param {string} customMetricName
     *   A fully-qualified path representing CustomMetric resource.
     * @returns {string} A string representing the custom_metric.
     */
    matchCustomMetricFromCustomMetricName(customMetricName) {
        return this.pathTemplates.customMetricPathTemplate.match(customMetricName).custom_metric;
    }
    /**
     * Return a fully-qualified dataRetentionSettings resource name string.
     *
     * @param {string} property
     * @returns {string} Resource name string.
     */
    dataRetentionSettingsPath(property) {
        return this.pathTemplates.dataRetentionSettingsPathTemplate.render({
            property: property,
        });
    }
    /**
     * Parse the property from DataRetentionSettings resource.
     *
     * @param {string} dataRetentionSettingsName
     *   A fully-qualified path representing DataRetentionSettings resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromDataRetentionSettingsName(dataRetentionSettingsName) {
        return this.pathTemplates.dataRetentionSettingsPathTemplate.match(dataRetentionSettingsName).property;
    }
    /**
     * Return a fully-qualified dataSharingSettings resource name string.
     *
     * @param {string} account
     * @returns {string} Resource name string.
     */
    dataSharingSettingsPath(account) {
        return this.pathTemplates.dataSharingSettingsPathTemplate.render({
            account: account,
        });
    }
    /**
     * Parse the account from DataSharingSettings resource.
     *
     * @param {string} dataSharingSettingsName
     *   A fully-qualified path representing DataSharingSettings resource.
     * @returns {string} A string representing the account.
     */
    matchAccountFromDataSharingSettingsName(dataSharingSettingsName) {
        return this.pathTemplates.dataSharingSettingsPathTemplate.match(dataSharingSettingsName).account;
    }
    /**
     * Return a fully-qualified dataStream resource name string.
     *
     * @param {string} property
     * @param {string} data_stream
     * @returns {string} Resource name string.
     */
    dataStreamPath(property, dataStream) {
        return this.pathTemplates.dataStreamPathTemplate.render({
            property: property,
            data_stream: dataStream,
        });
    }
    /**
     * Parse the property from DataStream resource.
     *
     * @param {string} dataStreamName
     *   A fully-qualified path representing DataStream resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromDataStreamName(dataStreamName) {
        return this.pathTemplates.dataStreamPathTemplate.match(dataStreamName).property;
    }
    /**
     * Parse the data_stream from DataStream resource.
     *
     * @param {string} dataStreamName
     *   A fully-qualified path representing DataStream resource.
     * @returns {string} A string representing the data_stream.
     */
    matchDataStreamFromDataStreamName(dataStreamName) {
        return this.pathTemplates.dataStreamPathTemplate.match(dataStreamName).data_stream;
    }
    /**
     * Return a fully-qualified firebaseLink resource name string.
     *
     * @param {string} property
     * @param {string} firebase_link
     * @returns {string} Resource name string.
     */
    firebaseLinkPath(property, firebaseLink) {
        return this.pathTemplates.firebaseLinkPathTemplate.render({
            property: property,
            firebase_link: firebaseLink,
        });
    }
    /**
     * Parse the property from FirebaseLink resource.
     *
     * @param {string} firebaseLinkName
     *   A fully-qualified path representing FirebaseLink resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromFirebaseLinkName(firebaseLinkName) {
        return this.pathTemplates.firebaseLinkPathTemplate.match(firebaseLinkName).property;
    }
    /**
     * Parse the firebase_link from FirebaseLink resource.
     *
     * @param {string} firebaseLinkName
     *   A fully-qualified path representing FirebaseLink resource.
     * @returns {string} A string representing the firebase_link.
     */
    matchFirebaseLinkFromFirebaseLinkName(firebaseLinkName) {
        return this.pathTemplates.firebaseLinkPathTemplate.match(firebaseLinkName).firebase_link;
    }
    /**
     * Return a fully-qualified googleAdsLink resource name string.
     *
     * @param {string} property
     * @param {string} google_ads_link
     * @returns {string} Resource name string.
     */
    googleAdsLinkPath(property, googleAdsLink) {
        return this.pathTemplates.googleAdsLinkPathTemplate.render({
            property: property,
            google_ads_link: googleAdsLink,
        });
    }
    /**
     * Parse the property from GoogleAdsLink resource.
     *
     * @param {string} googleAdsLinkName
     *   A fully-qualified path representing GoogleAdsLink resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromGoogleAdsLinkName(googleAdsLinkName) {
        return this.pathTemplates.googleAdsLinkPathTemplate.match(googleAdsLinkName).property;
    }
    /**
     * Parse the google_ads_link from GoogleAdsLink resource.
     *
     * @param {string} googleAdsLinkName
     *   A fully-qualified path representing GoogleAdsLink resource.
     * @returns {string} A string representing the google_ads_link.
     */
    matchGoogleAdsLinkFromGoogleAdsLinkName(googleAdsLinkName) {
        return this.pathTemplates.googleAdsLinkPathTemplate.match(googleAdsLinkName).google_ads_link;
    }
    /**
     * Return a fully-qualified keyEvent resource name string.
     *
     * @param {string} property
     * @param {string} key_event
     * @returns {string} Resource name string.
     */
    keyEventPath(property, keyEvent) {
        return this.pathTemplates.keyEventPathTemplate.render({
            property: property,
            key_event: keyEvent,
        });
    }
    /**
     * Parse the property from KeyEvent resource.
     *
     * @param {string} keyEventName
     *   A fully-qualified path representing KeyEvent resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromKeyEventName(keyEventName) {
        return this.pathTemplates.keyEventPathTemplate.match(keyEventName).property;
    }
    /**
     * Parse the key_event from KeyEvent resource.
     *
     * @param {string} keyEventName
     *   A fully-qualified path representing KeyEvent resource.
     * @returns {string} A string representing the key_event.
     */
    matchKeyEventFromKeyEventName(keyEventName) {
        return this.pathTemplates.keyEventPathTemplate.match(keyEventName).key_event;
    }
    /**
     * Return a fully-qualified measurementProtocolSecret resource name string.
     *
     * @param {string} property
     * @param {string} data_stream
     * @param {string} measurement_protocol_secret
     * @returns {string} Resource name string.
     */
    measurementProtocolSecretPath(property, dataStream, measurementProtocolSecret) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.render({
            property: property,
            data_stream: dataStream,
            measurement_protocol_secret: measurementProtocolSecret,
        });
    }
    /**
     * Parse the property from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).property;
    }
    /**
     * Parse the data_stream from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the data_stream.
     */
    matchDataStreamFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).data_stream;
    }
    /**
     * Parse the measurement_protocol_secret from MeasurementProtocolSecret resource.
     *
     * @param {string} measurementProtocolSecretName
     *   A fully-qualified path representing MeasurementProtocolSecret resource.
     * @returns {string} A string representing the measurement_protocol_secret.
     */
    matchMeasurementProtocolSecretFromMeasurementProtocolSecretName(measurementProtocolSecretName) {
        return this.pathTemplates.measurementProtocolSecretPathTemplate.match(measurementProtocolSecretName).measurement_protocol_secret;
    }
    /**
     * Return a fully-qualified property resource name string.
     *
     * @param {string} property
     * @returns {string} Resource name string.
     */
    propertyPath(property) {
        return this.pathTemplates.propertyPathTemplate.render({
            property: property,
        });
    }
    /**
     * Parse the property from Property resource.
     *
     * @param {string} propertyName
     *   A fully-qualified path representing Property resource.
     * @returns {string} A string representing the property.
     */
    matchPropertyFromPropertyName(propertyName) {
        return this.pathTemplates.propertyPathTemplate.match(propertyName).property;
    }
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns {Promise} A promise that resolves when the client is closed.
     */
    close() {
        if (this.analyticsAdminServiceStub && !this._terminated) {
            return this.analyticsAdminServiceStub.then(stub => {
                this._log.info('ending gRPC channel');
                this._terminated = true;
                stub.close();
            });
        }
        return Promise.resolve();
    }
}
exports.AnalyticsAdminServiceClient = AnalyticsAdminServiceClient;
